### ISSUE TITLE
Activity not launching/working in sugar-build.

### ISSUE THREAD

**Issue by i5o** [sugarlabs/browse-activity#38]
Using sugar-build (latest one! :P) I get

```
Traceback (most recent call last):
  File "/home/broot/sugar-build/build/out/install/bin/sugar-activity", line 220, in <module>
    main()
  File "/home/broot/sugar-build/build/out/install/bin/sugar-activity", line 164, in main
    module = __import__(module_name)
  File "/home/broot/sugar-build/build/out/install/share/sugar/activities/Browse.activity/webactivity.py", line 53, in <module>
    from widgets import TitledTray
  File "/home/broot/sugar-build/build/out/install/share/sugar/activities/Browse.activity/widgets.py", line 109, in <module>
    css_provider.load_from_data(css)
gi._glib.GError: <data>:7:14'min-height' is not a valid property name
Exited with status 1, pid 29998 data (<open file '<fdopen>', mode 'w' at 0xab81de0>, '8d463c472337246bd4f6134416a1aee8a667ea88')
```

Maybe it works on fedora rawhide, or smh. But we should provide support to sugar-build on debian or similars.

**Comment by quozl** [comment]
Yes, I got that too on Ubuntu 16.04, so I'm holding a packaging patch to fix it.

```
From 881452f90e722e23267860a73c7c5d12f66b37b6 Mon Sep 17 00:00:00 2001
From: James Cameron <quozl@laptop.org>
Date: Fri, 16 Sep 2016 12:05:31 +1000
Subject: [PATCH 1/3] Ubuntu 16.04 - fix fail to start

Ubuntu 16.04 has Gtk+ 3.18.9 which does not have min-height, causing
activity fail to start.

Ignore the failure.
---
 widgets.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/widgets.py b/widgets.py
index 8eaf62c..71e87c8 100644
--- a/widgets.py
+++ b/widgets.py
@@ -106,7 +106,10 @@ css = ('''
 '''.format(
     cell2over5=(style.GRID_CELL_SIZE*2)/5
 ))
-css_provider.load_from_data(css)
+try:
+    css_provider.load_from_data(css)
+except:
+    pass  # Gtk+ 3.18.9 does not have min-height
 context = Gtk.StyleContext()
 context.add_provider_for_screen(screen, css_provider,
                                 Gtk.STYLE_PROVIDER_PRIORITY_USER)

```

**Comment by i5o** [comment]
Does that breaks anything? I mean, the activity looks different or smh? If not, you should submit a pull request.

I will test the patch anyway

**Comment by i5o** [comment]
The activity works with that.

 but I think there is some problem with my computer or idk! WebKit seems to don't work on my laptop

Failed to create shared memory file /WK2SharedMemory.1340966597
Failed to create shared memory file /WK2SharedMemory.3276060909
```
Failed to create shared memory file /WK2SharedMemory.1084201496
Failed to create shared memory file /WK2SharedMemory.1232222103
Failed to create shared memory file /WK2SharedMemory.1544515977
Failed to create shared memory file /WK2SharedMemory.18382483
```

but ignoring that (again), it works

**Comment by i5o** [comment]
Tried to stop activity from button, got

```
Traceback (most recent call last):
  File "/home/broot/sugar-build/build/out/install/lib/python2.7/site-packages/sugar3/activity/activity.py", line 1087, in _prepare_close
    self.save()
  File "/home/broot/sugar-build/build/out/install/lib/python2.7/site-packages/sugar3/activity/activity.py", line 896, in save
    self.write_file(file_path)
  File "/home/broot/sugar-build/build/out/install/share/sugar/activities/Browse.activity/webactivity.py", line 373, in write_file
    self._tabbed_view.get_state()
  File "/home/broot/sugar-build/build/out/install/share/sugar/activities/Browse.activity/browser.py", line 487, in get_state
    'state': tab_page.browser.get_state()})
  File "/home/broot/sugar-build/build/out/install/share/sugar/activities/Browse.activity/browser.py", line 680, in get_state
    state = self.get_session_state()
AttributeError: 'Browser' object has no attribute 'get_session_state'

```

**Comment by quozl** [comment]
> If not, you should submit a pull request.

Perhaps, if I get time.  You're welcome to take it on, and commit with `--author`.

> Failed to create shared memory file

Yes, I've seen that, but haven't dug into it, as it doesn't break anything.

> AttributeError: 'Browser' object has no attribute 'get_session_state'

Yes, I've fixed that too, in packaging patches for Ubuntu 16.04, to fix a corner case where Browse would fail between when the laptop was turned on and the first security updates installed.  What version of WebKit are you using?

```
From a67dcfc5260ad83a055c061aad8036fbb35b9839 Mon Sep 17 00:00:00 2001
From: James Cameron <quozl@laptop.org>
Date: Fri, 16 Sep 2016 12:06:59 +1000
Subject: [PATCH 2/3] Ubuntu 16.04 - fix fail to stop

Ubuntu 16.04 was released with WebKit 2.10.9 which does not have
get_session_state(), causing keep error on stop.  Since, WebKit 2.12.5
was in security updates.

Ignore failure rather than make error.
---
 browser.py | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/browser.py b/browser.py
index e2d54a1..59f8faf 100644
--- a/browser.py
+++ b/browser.py
@@ -677,10 +677,13 @@ class Browser(WebKit2.WebView):
         self.cached_uri = None
 
     def get_state(self):
-        state = self.get_session_state()
-        gbytes = state.serialize()
-        # JSON results in utf8-decoding, so it needs to be good data
-        return b64encode(gbytes.get_data())
+        try:
+            state = self.get_session_state()
+            gbytes = state.serialize()
+            # JSON results in utf8-decoding, so it needs to be good data
+            return b64encode(gbytes.get_data())
+        except:
+            return ''  # graceful degradation in case of old WebKit2
 
     def get_legacy_history(self):
         """Return the browsing history of this browser."""
-- 
1.9.1

```

**Comment by i5o** [comment]
The one that sugar-build installs.

3.0

I'll test and submit a pull request (maybe tomorrow!)

Thanks James :D

2016-11-15 0:09 GMT-03:00, James Cameron <notifications@github.com>:
>> If not, you should submit a pull request.
>
> Perhaps, if I get time.  You're welcome to take it on, and commit with
> `--author`.
>
>> Failed to create shared memory file
>
> Yes, I've seen that, but haven't dug into it, as it doesn't break anything.
>
>> AttributeError: 'Browser' object has no attribute 'get_session_state'
>
> Yes, I've fixed that too, in packaging patches for Ubuntu 16.04, to fix a
> corner case where Browse would fail between when the laptop was turned on
> and the first security updates installed.  What version of WebKit are you
> using?
>
> ```
> From a67dcfc5260ad83a055c061aad8036fbb35b9839 Mon Sep 17 00:00:00 2001
> From: James Cameron <quozl@laptop.org>
> Date: Fri, 16 Sep 2016 12:06:59 +1000
> Subject: [PATCH 2/3] Ubuntu 16.04 - fix fail to stop
>
> Ubuntu 16.04 was released with WebKit 2.10.9 which does not have
> get_session_state(), causing keep error on stop.  Since, WebKit 2.12.5
> was in security updates.
>
> Ignore failure rather than make error.
> ---
>  browser.py | 11 +++++++----
>  1 file changed, 7 insertions(+), 4 deletions(-)
>
> diff --git a/browser.py b/browser.py
> index e2d54a1..59f8faf 100644
> --- a/browser.py
> +++ b/browser.py
> @@ -677,10 +677,13 @@ class Browser(WebKit2.WebView):
>          self.cached_uri = None
>
>      def get_state(self):
> -        state = self.get_session_state()
> -        gbytes = state.serialize()
> -        # JSON results in utf8-decoding, so it needs to be good data
> -        return b64encode(gbytes.get_data())
> +        try:
> +            state = self.get_session_state()
> +            gbytes = state.serialize()
> +            # JSON results in utf8-decoding, so it needs to be good data
> +            return b64encode(gbytes.get_data())
> +        except:
> +            return ''  # graceful degradation in case of old WebKit2
>
>      def get_legacy_history(self):
>          """Return the browsing history of this browser."""
> --
> 1.9.1
>
> ```
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/sugarlabs/browse-activity/issues/38#issuecomment-260535007


-- 
Ignacio Rodríguez
